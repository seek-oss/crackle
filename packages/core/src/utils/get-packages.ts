import fs from 'fs/promises';
import path from 'path';

import glob from 'fast-glob';

import type { EnhancedConfig } from '../config';

export interface Package {
  name: string;
  root: string;
}

export type Packages = Map<string, Package>;

export const getPackages = async (
  config: EnhancedConfig,
): Promise<Packages> => {
  const monorepoPackages = await glob(['packages/*/package.json'], {
    cwd: config.root,
    absolute: true,
  });

  const isMonorepo = monorepoPackages.length > 0;

  const allPackageJsons = isMonorepo
    ? monorepoPackages
    : [config.resolveFromRoot('package.json')];

  const packages = new Map<string, Package>();

  for (const packageJsonPath of allPackageJsons) {
    const packageJson = JSON.parse(
      await fs.readFile(packageJsonPath, { encoding: 'utf-8' }),
    );

    packages.set(packageJson.name, {
      name: packageJson.name,
      root: path.dirname(packageJsonPath),
    });
  }

  return packages;
};
