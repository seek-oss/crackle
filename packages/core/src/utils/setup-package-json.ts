import assert from 'assert';
import fs from 'fs';
import path from 'path';
import { isDeepStrictEqual } from 'util';

// @ts-ignore no declaration file
import structuredClonePolyfill from '@ungap/structured-clone';
import fse from 'fs-extra';

import { distDir, sideEffectsDir } from '../constants';
import type { PackageEntryPoint as Entry, PackageJson } from '../types';

import { writePackageJson } from './files';
import { partition } from './partition';
import { moduleHasSideEffects } from './side-effects';

export type Difference =
  | { key: 'main' | 'module' | 'types'; from?: string; to?: string }
  | { key: 'files'; additions: string[] }
  | { key: 'sideEffects'; additions: string[] }
  | { key: 'exports' }
  | { key: 'order' };

type ExportString = `./${string}`;
interface ExportObject {
  types: ExportString | Omit<ExportObject, 'types'>;
  import: ExportString;
  require: ExportString;
}
type Exports = Record<string, ExportString | ExportObject>;

const structuredClone = global.structuredClone ?? structuredClonePolyfill;

const sort = <T>(items: Iterable<T>, key?: keyof T) =>
  Array.from(items).sort((a, b) => {
    const aValue = key ? a[key] : a;
    const bValue = key ? b[key] : b;
    return aValue > bValue ? 1 : -1;
  });

const arrayDiff = <T>(array: T[], comparison: T[] | undefined) => {
  const existingItems = new Set(comparison ?? []);
  const missingItems = array.filter((file) => !existingItems.has(file));
  return missingItems;
};

const exportsDiff = (expected: Exports, actual: Exports) => {
  // Compare only keys which are the same in both objects.
  // This allows exports to be injected by `updatePackageJsonExports`
  const actualEntries = Object.entries(actual).filter(
    ([key]) => key in expected,
  );
  return !isDeepStrictEqual(Object.entries(expected), actualEntries);
};

const makeRelative = (value: string): ExportString =>
  value.startsWith('./') ? (value as ExportString) : `./${value}`;

const getExportsForPackage = (entries: Entry[], options: { from: string }) => {
  const [$default, other] = partition(entries, (entry) => entry.isDefaultEntry);
  const sortedEntries = [...$default, ...sort(other, 'entryName')];

  const exports: Exports = {};
  for (const entry of sortedEntries) {
    exports[entry.isDefaultEntry ? '.' : makeRelative(entry.entryName)] = {
      types: {
        import: makeRelative(entry.getOutputPath('dtsm', options)),
        require: makeRelative(entry.getOutputPath('dts', options)),
      },
      import: makeRelative(entry.getOutputPath('esm', options)),
      require: makeRelative(entry.getOutputPath('cjs', options)),
    };
  }
  exports[makeRelative('package.json')] = makeRelative('package.json');

  return exports as PackageJson['exports'];
};

const getSideEffectsForPackage = (
  entries: Entry[],
  existingSideEffects: PackageJson['sideEffects'],
  options: { from: string },
) => {
  if (
    typeof existingSideEffects === 'boolean' ||
    typeof existingSideEffects === 'undefined'
  ) {
    return {
      sideEffects: existingSideEffects,
      missingSideEffects: [],
    };
  }

  if (typeof existingSideEffects === 'string') {
    // eslint-disable-next-line no-param-reassign
    existingSideEffects = [existingSideEffects];
  }

  assert(
    Array.isArray(existingSideEffects),
    `sideEffects should be an array, but is ${existingSideEffects}.`,
  );

  const sideEffects = [...existingSideEffects];

  const addIfRequired = (flag: string) => {
    if (!sideEffects.includes(flag)) sideEffects.push(flag);
  };

  for (const entry of entries) {
    // we must use the relative path to ensure the package path doesn't match the flag
    const relativeEntryPath = path.relative(options.from, entry.entryPath);
    if (moduleHasSideEffects(relativeEntryPath, sideEffects)) {
      for (const outputPath of [
        entry.getOutputPath('cjs', options),
        entry.getOutputPath('esm', options),
      ]) {
        // add a new flag only if existing flags don't match
        if (!moduleHasSideEffects(outputPath, sideEffects)) {
          addIfRequired(outputPath.replace(path.extname(outputPath), '.*'));
        }
      }
    }
  }
  // catch-all flag
  addIfRequired(`${distDir}/${sideEffectsDir}/**`);

  return {
    sideEffects: sort(sideEffects),
    missingSideEffects: arrayDiff(sideEffects, existingSideEffects),
  };
};

export const diffPackageJson = async (
  packageRoot: string,
  packageJson: PackageJson,
  entries: Entry[],
): Promise<{
  diffs: Difference[];
  expectedPackageJson: PackageJson;
}> => {
  const diffs: Difference[] = [];

  // create expected package.json

  let expected = structuredClone(packageJson);

  const options = { from: packageRoot };
  expected.exports = getExportsForPackage(entries, options);

  const files = new Set(packageJson.files);
  for (const entry of entries) {
    if (entry.isDefaultEntry) {
      expected.main = makeRelative(entry.getOutputPath('cjs', options));
      expected.module = makeRelative(entry.getOutputPath('esm', options));
      expected.types = makeRelative(entry.getOutputPath('dts', options));
    }
    files.add(entry.entryName);
  }
  expected.files = sort(files);

  const { sideEffects, missingSideEffects } = getSideEffectsForPackage(
    entries,
    packageJson.sideEffects,
    options,
  );
  if (missingSideEffects.length) {
    expected.sideEffects = sideEffects;
  }

  const { default: sortPackageJson } = await import('sort-package-json');
  expected = sortPackageJson(expected);

  // do checks against expected package.json

  (['main', 'module', 'types'] as const).forEach((key) => {
    if (expected[key] !== packageJson[key]) {
      diffs.push({ key, from: packageJson[key], to: expected[key] });
    }
  });

  if (
    exportsDiff(
      expected.exports as Exports,
      (packageJson.exports ?? {}) as Exports,
    )
  ) {
    diffs.push({ key: 'exports' });
  }

  if (!isDeepStrictEqual(expected.files, packageJson.files)) {
    diffs.push({
      key: 'files',
      additions: arrayDiff(expected.files!, packageJson.files),
    });
  }

  if (!isDeepStrictEqual(expected.sideEffects, packageJson.sideEffects)) {
    diffs.push({ key: 'sideEffects', additions: missingSideEffects });
  }

  const expectedPackageJsonKeys = Object.keys(expected);
  const packageJsonKeys = Object.keys(packageJson);
  if (
    expectedPackageJsonKeys.length === packageJsonKeys.length &&
    !isDeepStrictEqual(expectedPackageJsonKeys, packageJsonKeys)
  ) {
    diffs.push({ key: 'order' });
  }

  return {
    diffs,
    expectedPackageJson: expected,
  };
};

const setupPackageJson =
  (write: boolean) =>
  async (packageRoot: string, entries: Entry[]): Promise<Difference[]> => {
    const packagePath = path.join(packageRoot, 'package.json');
    const packageJson: PackageJson = await fse.readJson(packagePath, { fs });

    const { diffs, expectedPackageJson } = await diffPackageJson(
      packageRoot,
      packageJson,
      entries,
    );

    if (write && diffs.length > 0) {
      await writePackageJson({
        dir: packageRoot,
        contents: expectedPackageJson,
      });
    }

    return diffs;
  };

export const validatePackageJson = setupPackageJson(false);
export const fixPackageJson = setupPackageJson(true);

export const updatePackageJsonExports = async (
  packageRoot: string,
  exports: string[],
) => {
  if (exports.length === 0) return;

  const packagePath = path.join(packageRoot, 'package.json');
  const packageJson: PackageJson = await fse.readJson(packagePath, { fs });

  const packageExports = packageJson.exports as Exports;
  const [lastKey, lastExport] = Object.entries(packageExports).pop()!;

  delete packageExports[lastKey];
  for (const entry of exports) {
    packageExports[makeRelative(entry)] = makeRelative(entry);
  }
  packageExports[lastKey] = lastExport;

  packageJson.exports = packageExports as PackageJson['exports'];

  await writePackageJson({
    dir: packageRoot,
    contents: packageJson,
  });
};
