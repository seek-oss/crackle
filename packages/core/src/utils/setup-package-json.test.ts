import fs from 'fs/promises';

import fixturez from 'fixturez';
import { describe, expect, test } from 'vitest';

import { getPackageEntryPoints } from './get-packages';
import { fixPackageJson, validatePackageJson } from './setup-package-json';

describe('setup-package-json', () => {
  const f = fixturez(__dirname);
  const readFile = (path: string) => fs.readFile(path, 'utf-8');

  test('validatePackageJson', async () => {
    const fixturePath = f.copy('library-with-docs');
    const entries = await getPackageEntryPoints({
      packageRoot: fixturePath,
    });

    const diffs = await validatePackageJson(fixturePath, entries);

    expect(diffs).toMatchSnapshot();
  });

  test('fixPackageJson', async () => {
    const fixturePath = f.copy('library-with-docs');
    const entries = await getPackageEntryPoints({
      packageRoot: fixturePath,
    });

    const pkgBefore = await readFile(`${fixturePath}/package.json`);
    await fixPackageJson(fixturePath, entries);
    const pkgAfter = await readFile(`${fixturePath}/package.json`);

    expect({
      diffA: pkgBefore,
      diffB: pkgAfter,
    }).toMatchSnapshot();
  });
});
