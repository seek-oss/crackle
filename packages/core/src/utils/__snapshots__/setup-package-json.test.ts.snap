// Vitest Snapshot v1

exports[`diffPackageJson > Empty package.json > diffs 1`] = `
[
  {
    "from": undefined,
    "key": "main",
    "to": "./dist/index.cjs",
  },
  {
    "from": undefined,
    "key": "module",
    "to": "./dist/index.esm.js",
  },
  {
    "additions": [
      "/dist",
      "/css",
      "/themes/apac",
    ],
    "key": "files",
  },
  {
    "key": "exports",
  },
]
`;

exports[`diffPackageJson > Empty package.json > expectedPackageJson 1`] = `
{
  "exports": {
    ".": {
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs",
      "types": "./dist/index.cjs.d.ts",
    },
    "./css": {
      "import": "./css/dist/index.esm.js",
      "require": "./css/dist/index.cjs",
      "types": "./css/dist/index.cjs.d.ts",
    },
    "./package.json": "./package.json",
    "./themes/apac": {
      "import": "./themes/apac/dist/index.esm.js",
      "require": "./themes/apac/dist/index.cjs",
      "types": "./themes/apac/dist/index.cjs.d.ts",
    },
  },
  "files": [
    "/css",
    "/dist",
    "/themes/apac",
  ],
  "main": "./dist/index.cjs",
  "module": "./dist/index.esm.js",
}
`;

exports[`diffPackageJson > Incorrect package.json > diffs 1`] = `
[
  {
    "from": "dist/index.esm.invalid",
    "key": "module",
    "to": "./dist/index.esm.js",
  },
  {
    "additions": [
      "/themes/apac",
    ],
    "key": "files",
  },
  {
    "key": "exports",
  },
]
`;

exports[`diffPackageJson > Incorrect package.json > expectedPackageJson 1`] = `
{
  "exports": {
    ".": {
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs",
      "types": "./dist/index.cjs.d.ts",
    },
    "./css": {
      "import": "./css/dist/index.esm.js",
      "require": "./css/dist/index.cjs",
      "types": "./css/dist/index.cjs.d.ts",
    },
    "./package.json": "./package.json",
    "./themes/apac": {
      "import": "./themes/apac/dist/index.esm.js",
      "require": "./themes/apac/dist/index.cjs",
      "types": "./themes/apac/dist/index.cjs.d.ts",
    },
  },
  "files": [
    "/apac",
    "/css",
    "/dist",
    "/themes/apac",
  ],
  "main": "./dist/index.cjs",
  "module": "./dist/index.esm.js",
}
`;
